### PR TITLE
[bitnami/ghost] Add Ghost Helm Chart tests

### PR DESCRIPTION
**Pull request by joancafom**
**Description of the change**

**Test results**

9 consecutive executions without issues: [Batch #1](https://github.com/joancafom/charts/runs/5980673253?check_suite_focus=true), [Batch #2](https://github.com/joancafom/charts/runs/5985970141?check_suite_focus=true) & [Batch #3](https://github.com/joancafom/charts/runs/5978684886?check_suite_focus=true).

Samples:

- [GKE](https://cp.production.k.bromelia.bitnami.net/v1/execution-graphs/606ebebf-15d3-4007-94aa-68aea2f1efa9/report) - PASS
- [AKS](https://cp.production.k.bromelia.bitnami.net/v1/execution-graphs/11f53449-f1ff-4c33-be6d-07531e97d7cb/report) - PASS
- [TKG](https://cp.production.k.bromelia.bitnami.net/v1/execution-graphs/0d833092-3a2a-4408-a8f6-4bee94bf59d0/report) - PASS

Adding the tests for the Ghost Helm chart. 

**Benefits**

When someone makes a change in the Helm Chart, they will be able to check if their change has broken the Helm chart by running these automated tests.

**Possible drawbacks**

With UI tests, there is always a possibility for flakiness. 

**Checklist**

- [X] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/master/CONTRIBUTING.md#sign-your-work)

### CODE REVIEW

File: .github/workflows/vib-ghost.yaml [hunk 1]
```diff
@@ -0,0 +1,36 @@
+name: 'vib'
+on: # rebuild any PRs and main branch changes
```
Comment (superaleks) [inline .github/workflows/vib-ghost.yaml:2]: I already have the PR ready for fixing inconsistencies. Ghost will **not** be included, so there is a need to make the inconsistency related changes here.

File: .github/workflows/vib-ghost.yaml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  CSP_API_TOKEN: ${{ secrets.CSP_API_TOKEN }}
+  VIB_PUBLIC_URL: https://cp.bromelia.vmware.com
+jobs:
+  vib-k8s-verify: # verify in multiple target platforms
+    runs-on: ubuntu-latest
+    environment: vmware-image-builder
+    strategy:
+      matrix:
+       include:
+         - target-platform: gke
```
Comment (superaleks) [inline .github/workflows/vib-ghost.yaml:19]:
```suggestion
         - target-platform: Tanzu Kubernetes Grid
```

File: .github/workflows/vib-ghost.yaml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+           target-platform-id: 91d398a2-25c4-4cda-8732-75a3cfc179a1
+         - target-platform: aks
+           target-platform-id: 7b13a7bb-011c-474f-ad71-8152fc321b9e
+         - target-platform: tkg
+           target-platform-id: 7ddab896-2e4e-4d58-a501-f79897eba3a0
+      fail-fast: false
+    name: K8s verify ${{ matrix.target-platform}}
```
Comment (superaleks) [inline .github/workflows/vib-ghost.yaml:26]:
```suggestion
    name: Ghost on ${{ matrix.target-platform }}
```

File: .github/workflows/vib-ghost.yaml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          ref: ${{github.event.pull_request.head.ref}}
+          repository: ${{github.event.pull_request.head.repo.full_name}}
```
Comment (superaleks) [inline .github/workflows/vib-ghost.yaml:31]:
```suggestion
          repository: ${{ github.event.pull_request.head.repo.full_name }}
```

File: .vib/ghost/cypress/cypress/support/commands.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    const origVal = originalFn(...args);
+
+    return new Promise((resolve) => {
+      setTimeout(() => {
+        resolve(origVal);
+      }, COMMAND_DELAY);
+    });
+  });
+}
+
```
Comment (superaleks) [inline .vib/ghost/cypress/cypress/support/commands.js:15]: I appreciate the necessity to explain various corner case behaviours, but I'd prefer to keep this data documented internally instead of having them in the comments. Let's create a task for this in our Helm chart strategy and brainstorm different ways of documenting this. 
The cognitive complexity of the code itself is already there for the reader, it is quite difficult to cope with if we add a lot of comments instead.

File: .vib/ghost/cypress/cypress/integration/ghost_spec.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
+/// <reference types="cypress" />
+import { random, getPageUrlFromTitle, getUserFromEmail } from './utils';
+
+it('allows to log in and out', () => {
+  cy.login(Cypress.env('email'), Cypress.env('password'), false);
```
Comment (superaleks) [inline .vib/ghost/cypress/cypress/integration/ghost_spec.js:5]: Could we always have this set the 'waitForDashboard' to `true`?
We are waiting for it in the login test and in the `commands.js` default settings. 
Better said, In which case will we will **not** wait for the dashboard?

File: .vib/ghost/cypress/cypress/support/commands.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  (
+    email = Cypress.env('email'),
+    password = Cypress.env('password'),
+    waitForDashboard = true
+  ) => {
+    cy.visit('/ghost/');
+    cy.get('input[type="email"]').should('be.enabled').type(email);
+    cy.get('input[type="password"]').should('be.enabled').type(password);
+    cy.contains('button', 'Sign in').click();
+    if (waitForDashboard) {
```
Comment (superaleks) [inline .vib/ghost/cypress/cypress/support/commands.js:36]: Same question as in the comment above, can you clarify why we need the `waitForDashboard`?

File: .vib/ghost/cypress/cypress.json [hunk 8]
```diff
@@ -0,0 +1,10 @@
+{
+    "env": {
+      "email": "user@example.com",
+      "password": "Complicated123!4"
+    },
+    "hosts": {
+      "bitnami-ghost.my": "{{ TARGET_IP }}"
```
Comment (superaleks) [inline .vib/ghost/cypress/cypress.json:7]: Very nice!

File: .vib/ghost/cypress/cypress/integration/ghost_spec.js [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+  cy.contains('Retry').should('not.exist');
+  // Dashboard is hidden in the UI until correctly logged in
+  cy.contains('Dashboard').should('be.visible');
+
+  cy.get("div[class*='avatar']").click();
+  cy.contains('Sign out').click();
+  cy.contains('Sign in');
+});
+
+it('allows to create a new post', () => {
```
Comment (superaleks) [inline .vib/ghost/cypress/cypress/integration/ghost_spec.js:15]:
```suggestion
it('allows to create and publish a new post', () => {
```

File: .vib/ghost/goss/goss.yaml [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+    exec: which ghost
+    exit-status: 0
+    stdout:
+      - "/opt/bitnami/ghost/bin/ghost"
+    stderr: [ ]
+    timeout: 40000
+  ghost-version-ok:
+    exec: ghost version | grep --quiet "$APP_VERSION"
+    exit-status: 0
+    stdout: [ ]
```
Comment (superaleks) [inline .vib/ghost/goss/goss.yaml:12]: Why do we need the `stdout` here, if we are not verifying the output?

File: .vib/ghost/cypress/cypress/support/index.js [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+// You can change the location of this file or turn off
+// automatically serving support files with the
+// 'supportFile' configuration option.
+//
+// You can read more here:
+// https://on.cypress.io/configuration
+// ***********************************************************
+
+// Import commands.js using ES2015 syntax:
+import './commands'
```
Comment (FraPazGal) [inline .vib/ghost/cypress/cypress/support/index.js:17]:
```suggestion
import './commands';
```

File: .vib/ghost/cypress/cypress/integration/ghost_spec.js [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+  cy.login();
+  cy.visit('/ghost/#/members/import');
+
+  // The element exists in the UI while loading, but we need to
+  // wait until it is visible to proceed
+  cy.contains('h1', 'Import members').should('be.visible');
+  cy.get('input[type="file"]').selectFile('cypress/fixtures/members.csv', {
+    force: true,
+  });
+  cy.contains('button', 'Import').click();
```
Comment (FraPazGal) [inline .vib/ghost/cypress/cypress/integration/ghost_spec.js:89]: We are working with non-random data. Let's say the test fails checking the added members because temporarily the page doesn't load. Could we encounter an issue here if we need to retry the test?

**Review by FraPazGal — COMMENTED** [review]
The PR looks good! Could you please check my comments?

File: .github/workflows/vib-ghost.yaml (hunk 1 above)
Comment (joancafom) [reply .github/workflows/vib-ghost.yaml:2]: This does no longer apply after https://github.com/bitnami/charts/pull/9780. Marking this as resolved.

File: .vib/ghost/cypress/cypress.json (hunk 8 above)
Comment (joancafom) [reply .vib/ghost/cypress/cypress.json:7]: Thanks!

File: .vib/ghost/cypress/cypress/integration/ghost_spec.js [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+    cy.visit(getPageUrlFromTitle(`${$pages.newPage.title}-${random}`));
+    cy.contains(`${$pages.newPage.title}-${random}`);
+    cy.contains(`${$pages.newPage.content}-${random}`);
+  });
+});
+
+it('allows to import members', () => {
+  cy.login();
+  cy.visit('/ghost/#/members/import');
+
```
[the remark this answers is not in the document]
Comment (joancafom) [reply .vib/ghost/cypress/cypress/integration/ghost_spec.js:79]: Okay, agreed. If there is an explicit agreement on trying to avoid `be.visible/exists` when possible we should trust the author, and hence not need these kinds of comments.

File: .vib/ghost/cypress/cypress/integration/ghost_spec.js (hunk 12 above)
Comment (joancafom) [reply .vib/ghost/cypress/cypress/integration/ghost_spec.js:84]: That's a good point. Another good valid reasoning is that we want to execute tests twice in the future, maybe after scaling up/down replica sets.

In any case, Ghost's import members function does actually overwrite the data if it already existed, so it should not be a problem. In fact, I had manually tested that the tests can be run multiple times on the same deployment 😁

File: .vib/ghost/vib-platform-verify.json [hunk 14]
```diff
@@ -0,0 +1,66 @@
+{
```
[the remark this answers is not in the document]
Comment (joancafom) [reply .vib/ghost/vib-platform-verify.json:1]: Does no longer apply, I will resolve this thread.

File: .vib/ghost/cypress/cypress/integration/ghost_spec.js (hunk 6 above)
Comment (joancafom) [reply .vib/ghost/cypress/cypress/integration/ghost_spec.js:5]: > Better said, In which case will we will not wait for the dashboard?

This is the only case in which we **can't wait for it**. Upon submitting the login form, Ghost will do different redirections and eventually show the Dashboard page. It is not only until those last redirections happen that one can navigate to other pages, else you are presented with an Unauthorized error. In short, this is the reason why I decided to add it to commands.

Why don't we want to wait for it here? As we are validating the log-in feature, I want to explicitly validate that there are no form errors. If I wait for Dashboard to appear, it means that the log-in form is already gone.

I don't know if this is kind of tricky and I should just always validate that in `commands` and that's it. WDYT?

File: .vib/ghost/cypress/cypress/support/commands.js (hunk 5 above)
Comment (joancafom) [reply .vib/ghost/cypress/cypress/support/commands.js:15]: While I agree that it is good having this documented in our strategy, I still see it valuable to have these corner cases in the specific tests where it applies. If this is a workaround that most assets will use then I don't see the point either, but I believe that we won't have that many using `hosts`. In fact, this is the only one that currently has it out of all the others already implemented.

IMHO, I'd rather be too verbose in these workarounds in the code than not commenting at all. Someone that does not have the context will probably don't understand why I am overwriting the built-in `visit` function here.

File: .vib/ghost/cypress/cypress/support/commands.js (hunk 7 above)
Comment (joancafom) [reply .vib/ghost/cypress/cypress/support/commands.js:36]: Replied in the thread above. Also, there is a summary in the comments below.

File: .vib/ghost/goss/goss.yaml (hunk 10 above)
Comment (joancafom) [reply .vib/ghost/goss/goss.yaml:12]: I reckon this in fact validates that the output is empty. I mean, if there were an error and it was printed in `stdout`, this will fail as the output contains text.

File: .vib/ghost/cypress/cypress/integration/ghost_spec.js [hunk 15]
```diff
@@ -0,0 +1,116 @@
+/// <reference types="cypress" />
+import { random, getPageUrlFromTitle, getUserFromEmail } from './utils';
+
+it('allows to log in and out', () => {
+  cy.login();
+
```
Comment (superaleks) [inline .vib/ghost/cypress/cypress/integration/ghost_spec.js:6]: New line here, is it on purpose?

File: .vib/ghost/cypress/cypress/integration/ghost_spec.js [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+  cy.get("div[class*='avatar']").click();
+  cy.contains('Sign out').click();
+  cy.contains('Sign in');
+});
+
+it('allows to create and publish a new post', () => {
+  cy.login();
+  cy.visit('/ghost/#/posts');
+
```
Comment (superaleks) [inline .vib/ghost/cypress/cypress/integration/ghost_spec.js:15]: Same as above, new line.

File: .vib/ghost/cypress/cypress/integration/ghost_spec.js (hunk 15 above)
Comment (joancafom) [reply .vib/ghost/cypress/cypress/integration/ghost_spec.js]: I removed new lines 😁  thanks!

**Review by superaleks — APPROVED** [review]
LGTM!

**Review by FraPazGal — APPROVED** [review]
